File: ex07/c_version/ex7.c
/*********************************************************************
* CSCI203 - Exercise 7
* ex7.c - main() driver for implementation of Karp-Rabin string search algorithm
* Author: Dinh Che (codeninja55) | andrew at codeninja55.me
* UOW Details: Dinh Che (5721970) | dbac496 at uowmail.edu.au
* Last modified: 2018.09.24
*********************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include "helper.h"

#define BUFFER_SZ 100
#define G_SZ 25

typedef struct {
    int head, tail, capacity;
    int items[G_SZ];
} queue;

int n_size;

void print_matrix(int G[n_size][n_size]);
// Queue prototypes
queue* create_queue(int);
int is_full(queue*);
int is_empty(queue*);
void enqueue(int, queue*);
int dequeue(queue*);
void print_queue(queue *q);
// Graph prototypes
void init_matrix(int M[n_size][n_size]);
void set_edge(int G[n_size][n_size], int f, int t);
void BFS(int s, int adj[n_size][n_size], int v[]);


int main(int argc, const char * argv[])
{
    printf("|==========| CSCI203 Exercise 07 |==========|\n"
           "Author: Dinh Che (dbac496)\nRun at: %s", get_current_time());
    FILE *fd;
    char filename[BUFFER_SZ];

    if (argc == 2)
        strncpy(filename, argv[1], BUFFER_SZ);
    else {
        printf("Input file name >> ");
        fgets(filename, BUFFER_SZ, stdin);  // safe read
    }

    if ((fd = fopen(filename, "r")) == NULL) {
        printf("[!!] Error: Failed to open file\nExit(1)...\n");
        return -1;
    }

    fscanf(fd, "%d", &n_size);

    int G[n_size][n_size];
    int visited[G_SZ] = {0};  // TODO: Remember to remove variable size object
    init_matrix(G);

    int from, to;

    while (fscanf(fd, "%d %d", &from, &to) == 2)
        set_edge(G, from, to);

    fclose(fd);
    print_matrix(G);

    BFS(0, G, visited);

    int i;
    printf("Visited: ");
    for (i = 0; i < n_size; i++)
        printf("%d ", visited[i]);

    return 0;
}

void BFS(int s, int adj[n_size][n_size], int *v)
{
    int current, ctr, j, k;
    ctr = 0;
    queue *q = create_queue(n_size);

    enqueue(s, q);

    while (is_empty(q)) {
        print_queue(q);
        current = dequeue(q);
        printf("C: %d \n", current);
        v[ctr++] = current;

        for (j = 0; j < n_size; j++) {
            if (adj[current][j] == 1) {
                printf("G[%d][%d]: %d \n", current, j, adj[current][j]);
                bool found = false;
                for (k = 0; k < n_size; k++) {
                    if (j == v[k])
                        found = true;
                }
                if (found)
                    enqueue(j, q);
            }
        }
    }
}

queue* create_queue(int size)
{
    queue* q = ec_malloc(sizeof(queue));
    q->head = -1;
    q->tail = -1;
    q->capacity = size;
    return q;
}

int is_full(queue *q) { return (q->tail == q->capacity - 1) ? 1 : 0; }

int is_empty(queue* q) { return (q->tail == -1) ? 1 : 0; }

void enqueue(int elem, queue *q)
{
    if (is_full(q) == 1) {
        printf("Queue is full");
        return;
    } else {
        if (q->head == -1)
            q->head = 0;

        q->tail++;
        q->items[q->tail] = elem;
    }
}

int dequeue(queue *q)
{
    int ret;
    if (is_empty(q)) {
        printf("Queue is empty");
        ret = -1;
    } else {
        ret = q->items[q->head];
        q->head++;
        if (q->head > q->tail)
            q->head = q->tail = -1;
    }
    return ret;
}

void print_queue(queue *q)
{
    int i;
    if (is_empty(q)) {
        printf("Queue is empty");
    } else {
        printf("\nQueue contains \n");
        for(i = q->head; i < q->tail + 1; i++)
            printf("%d ", q->items[i]);
        printf("\n");
    }
}

void set_edge(int G[n_size][n_size], int f, int t)
{
    G[f][t] = 1;
    G[t][f] = 1;
}

void init_matrix(int M[n_size][n_size])
{
    int i, j;
    for (i = 0; i < n_size; i++) {
        for (j = 0; j < n_size; j++)
            M[i][j] = 0;
    }
}

void print_matrix(int G[n_size][n_size])
{
    int i, j;
    printf("     ");
    for (i = 0; i < n_size; i++) printf("%3d,", i);
    printf("\n");
    for (i = 0; i < n_size; i++) {
        printf("%2d  [", i);
        for (j = 0; j < n_size; j++) printf("%3d,", G[i][j]);
        printf("  ]\n");
    }
    printf("\n");
}

File: ex07/py_version/ex7.py
#######################################################################
# Title: CSCI203 - Exercise 7
# File: ex7.py - main() for exercise 6 execution
# Author: Dinh Che (codeninja55) | andrew at codeninja55.me
# UOW Details: dbac496 | 5721970 | dbac496 at uowmail.edu.au
# Last modified: 2018.09.24
#######################################################################
import os
import sys
from argparse import ArgumentParser
from pathlib import Path
from typing import List


# Globals
n_size = 0


class Queue:
    """Class Queue to represent an array implementation of a FIFO queue."""

    def __init__(self, capacity):
        self._size = 0
        self._head = -1
        self._tail = -1
        self._capacity = capacity
        self.Q = [None] * capacity  # initialise Q list

    # queue is full when size becomes equal to the capacity
    def is_full(self):
        return self._size == self._capacity

    # queue is empty when size is 0
    def is_empty(self):
        print("Queue is Empty.")
        return self._size == 0

    def enqueue(self, elem):
        if self.is_full():
            print("FULL")
            return

        if self._head is -1:
            self._head = 0
        self._tail += 1
        self.Q[self._tail] = elem
        self._size += 1

    def dequeue(self):
        if not self.is_empty:
            print("EMPTY")
            return

        ret = self.Q[self._head]
        self._head += 1
        if self._head > self._tail:
            self._head = self._tail = -1
        self._size -= 1
        return ret

    def front(self):
        if self.is_empty():
            print("Queue is empty")

        return self.Q[self._head]

    def rear(self):
        if self.is_empty():
            print("Queue is empty")

        return self.Q[self._tail]


class Graph:
    """A simple representation of graph using Adjacency Matrix."""

    def __init__(self, n_vertex):
        self.adj_matrix = [[0] * n_vertex for _ in range(n_vertex)]  # will create adj_matrix[[-1] * n_vertex]
        self._n_vertex = n_vertex
        self.vertices = {}
        self.vertices_list = [0] * n_vertex

    def set_vertex(self, idx):
        self.vertices[idx] = idx
        self.vertices_list[idx] = idx

    def set_edge(self, frm, to):
        frm = self.vertices[frm]
        to = self.vertices[to]
        self.adj_matrix[frm][to] = 1
        self.adj_matrix[to][frm] = 1  # for directed graph do not add this

    def get_vertex(self):
        return self.vertices_list

    def get_edges(self):
        edges = []
        for i in range(self._n_vertex):
            for j in range(self._n_vertex):
                if self.adj_matrix[i][j] is not 0:
                    edges.append((self.vertices_list[i], self.vertices_list[j]))
        return edges

    def get_matrix(self):
        return self.adj_matrix

    def size(self):
        return self._n_vertex


def BFS(s: any, adj: Graph) -> List:
    visited = []  # set of vertices visited
    q = Queue(adj.size())  # queue of vertex

    q.enqueue(s)

    while not q.is_empty():
        current = q.dequeue()
        visited.append(current)

        for v_idx, val in enumerate(adj.get_matrix()[current]):
            if v_idx not in visited and val is 1:
                q.enqueue(v_idx)
                print("Visited: " + str(v_idx))
    return visited


def read_file(filename):
    """
    Reads the file with the passed in filename and completes the implementation of karp_rabin_search().
    :param filename: char array[] string, filename of the file to read.
    """
    global n_size

    if filename:
        with open(filename, 'r') as fin:
            if fin.closed:
                sys.exit(1)

            # complete implementation below
            n_size = fin.readline()
            G = Graph(int(n_size))

            for line in fin.readlines():
                line_ls = line.split()
                v = int(line_ls[0])
                v_to = int(line_ls[1])
                G.set_vertex(v)
                G.set_edge(v, v_to)

            return G


def parse_args():
    """
    Wrapper function for argparse module to define the arguments for this script.
    :return: a parser with the arguments.
    """
    parser = ArgumentParser(prog='CN55 Exercise 7')
    parser.add_argument('filename', help='Enter file name to read from...')
    parser.add_argument('-f', '--filename', required=False, dest='filename')
    return parser.parse_args()


if __name__ == '__main__':
    print(os.getcwd())

    args = parse_args()
    if args.filename:
        file_path = Path(os.getcwd()) / args.filename
        G = read_file(filename=file_path)

    print("Number of Vertices: " + str(G.size()))
    print("Vertices of Graph")
    print(G.get_vertex())
    print("Edges of Graph")
    print(G.get_edges())
    print("Adjacency Matrix of Graph")
    print(G.get_matrix())

    # v = BFS(0, G)
    # print(v)
